### DISCUSSION TITLE
different results when forwarding audio in a batch

### DISCUSSION THREAD

**Discussion by jasonppy** [openai/whisper#381] [Q&A] [answered]
Hi,

Thanks for open sourcing the model and consistent support!

I find that when I decode audios in a batch, the result is different from decoding them one by one. The model is already set at .eval(). And as far as I know, layernorm doesn't use batch information in eval() phase. What do you think lead to this phenomenon?

Thanks in advance!

**Comment by jongwook** [comment]
Please see #81

**Reply by jasonppy** [reply]
Thanks for the quick reply!

I should've provide more context --- I'm using 
```python
results = whisper.decode(model, input_mel, options)
```
As the audio lengths are all shorter than 30sec.
Also I'm using Greedy Decoding.

**Reply by jongwook** [reply]
Thanks for the context! If the `options` have `temperature=0`, the model should behave the same in theory. But it's possible that the underlying CUDA implementations are different depending on the input shape, and in turn result in a slightly different outputs numerically. Are they significantly different or just slightly?

**Reply by jasonppy** [reply]
I do have `temperature=0` in options. The batch processing result is higher by ~2% in WER, which is significant

**Comment by jianfch** [comment, marked as answer]
I couldn't reproduce the phenomenon with the following code:
```python
import whisper
import torch

audio_paths = ['aud0.mp3', 'aud1.mp3', 'aud2.mp3', ...]
model = whisper.load_model('base')
options = whisper.DecodingOptions(language='en')
mels = [whisper.pad_or_trim(whisper.log_mel_spectrogram(path)[None], 3000).to(model.device) for path in audio_paths]
with torch.no_grad():
    results = [whisper.decode(model, mel, options) for mel in mels]
    batch_results = whisper.decode(model, torch.cat(mels), options)
assert set(i[0].text == j.text for i, j in zip(results, batch_results)) == {True}, 'text do not match'
```
give this a try to see if it changes anything

**Reply by jasonppy** [reply]
Thanks!

I made a [Colab notebook](https://colab.research.google.com/drive/1iDedl7Qd75xlXK_cYcNxPQQO5_jkrMxt?usp=sharing) using your code to show the inconsistency between sequential processing and batch processing. I'd really appreciate it if you or other people can take a look

**Reply by jongwook** [reply]
It didn't happen when I tried:

<img width="885" alt="image" src="https://user-images.githubusercontent.com/266841/197363083-bdaff55b-b7ac-408c-8779-57bf8677d45c.png">

It does seem possible that it's the slight differences between CUDA implementations for different shapes. Chinese performance of the tiny model is pretty bad anyway; do you see the same problem with medium or large models?

**Reply by jasonppy** [reply]
Thanks for your reply!

It happened again when I ran it just now
![image](https://user-images.githubusercontent.com/47729801/197363306-bdf63030-d64d-4a80-86a8-ae5c5b65c181.png)

When I switched medium or large in this notebook, I didn't see this problem in several trials of this notebook. However, in my larger scale experiments, I used the medium model and larger batch size, it the inconsistency always happen. It leads to non-trivial differences in WER.

But with the same model, same batch size, different random seeds always give the same WER.

**Reply by jianfch** [reply]
There are batch matrix operation optimizations that pytorch does in the background. This is why you can end up with slight numerical differences between a batch operation and non-batch operation with the same inputs. You can confirm this with a test like this:
```python
mels_batch = torch.cat(mels)
rand_weights = torch.rand_like(mels[0]).permute(0, 2, 1)
res = mels[0] @ rand_weights
res_non_batch = mels[0][0] @ rand_weights[0]
res_batch0 = mels_batch @ rand_weights
res_batch1 = mels_batch[:2] @ rand_weights

#check how many numbers are equal
equal_percent_nb = (res[0] == res_non_batch).float().mean().item() * 100
equal_percent = (res == res_batch0[0]).float().mean().item() * 100
equal_percent1 = (res_batch0[0] == res_batch1[0]).float().mean().item() * 100

#check how many numbers are close within specified tolerance(s)
close_percent = torch.isclose(res, res_batch0[0], rtol=1e-04).float().mean().item() * 100

print(f'Equal between non-batch and batch size 1: {equal_percent_nb}%\nEqual between batch size 1 and >1: {equal_percent:.04f}%\nEqual between batch size 2 and >2: {equal_percent1:.04f}%\nClose between batch size 1 and >1: {close_percent:.04f}%')

#Equal between non-batch and batch size 1: 100.0%
#Equal between batch size 1 and >1: 3.671874850988388%
#Equal between batch size 2 and >2: 100.0%
#Close between batch size 1 and >1: 99.96874928474426%
```
The slight differences are further amplified by the how there are two different languages spoken in the batch. That's why you will see slight difference in the decoded results too.
